components/automate-deployment/pkg/converge/converger.go

```diff
@@ -127,63 +136,460 @@ func StartConverger(opts ...ConvergerOpt) Converger {
+type stopRequest struct {
+	eventSink    EventSink
+	desiredState DesiredState
+	doneChan     chan error
+}
+
+func (s *stopRequest) SendResponse(err error) {
+	s.doneChan <- err
+	close(s.doneChan)
+}
+
+func (s *stopRequest) String() string { return "stopRequest" }
+
+// reconfigureDone is a user-sendable message that instructs the
+// converger that we've completed a reconfiguration request.
+type reconfigureDone struct {
+	doneChan chan error
+}
+
+func (r *reconfigureDone) SendResponse(err error) {
+	if r.doneChan != nil {
+		r.doneChan <- err
+		close(r.doneChan)
+	}
+}
+
+func (r *reconfigureDone) String() string { return "reconfigureDone" }
+
+// States
+//
+// We use the following state machine to handle the two "waiting"
+// states that we need to track inside the converger.
+//
+//        +-----------+  convergeRequest           +-------------------+
+//        |           |  err = ErrRestartPending   |                   |
+//        |           +--------------------------->|                   |
+//        |    idle   |                            | waitingForRestart |
+//        |           |<---------------------------+                   |
+//        |           |       timeout              |                   |
+//        +-------+---+                            +-------------------+
+//            ^   |
+//            |   | convergeRequest
+//   timeout 2|   | err = ErrSelfReconfigurePending
+//            |   |
+//            |   v
+//         +--+--------------------+
+//         |                       | timeout 1
+//         |                       +---------+
+//         | waitingForReconfigure |         |
+//         |                       |<--------+
+//         |                       |
+//         +-----------------------+
+//
+
+// idle is the default state of the converger. In this state we
+// are ready to accept any user-sendable requests. The idle state
+// never times out.
+type idle struct{}
+
+func (i *idle) ProcessMessage(c *converger, msg message) (state, error) {
+	switch msg.(type) {
+	case *convergeRequest:
+		c := c.compiler
+		req := msg.(*convergeRequest)
+		convergePlan, err := c.Compile(req.desiredState)
+		if err != nil {
+			logrus.WithError(err).Error("Failed to compile")
+			return i, err
+		}
+		err = convergePlan.Execute(req.eventSink)
+		return i.handleError(err)
+	case *stopRequest:
+		c := c.compiler
+		req := msg.(*stopRequest)
+		convergePlan, err := c.Compile(req.desiredState)
+		if err != nil {
+			logrus.WithError(err).Error("Failed to compile")
+			return i, err
+		}
+		err = convergePlan.Execute(req.eventSink)
+		return i.handleError(err)
+	case *reconfigureDone:
+		logrus.Warn("received reconfigureDone when in idle state. We should have been in waitingForReconfigure")
+		return i, nil
+	case *timeout:
+		logrus.Warn("received timeout when in idle state.")
+		return i, nil
+	default:
+		return i, UnknownMessageError(msg, i)
+	}
+}
+
+func (i *idle) handleError(err error) (state, error) {
```

Comment:
this isn't your fault, but it annoys me we use errors this way. maybe one day compiler(we should also rename that) could just return the next state or a real error?

Reply:
I agree it feels wrong that these are "errors" since these are expected outcomes of certain operations.